spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/bigquant/Linear.scala

```diff
@@ -0,0 +1,267 @@
+
+  val FAULT_TOLERANCE = 0.5f
+  val WEIGHT_THRESHOLD = 64.0f
+  val THRESHOLD = 127.0f
+
+  val empty: Tensor[T] = Tensor[T](1)
+
+  @transient
+  var _init = false
+
+  private def setWeightSum(weight: Tensor[T], weightSum: Array[T]): Unit = {
+    for (i <- 1 to outputSize) {
+      val singleRow = weight.select(1, i)
+      weightSum(i - 1) = singleRow.sum()
+    }
+  }
+
+  def initWeightAndBias(weightFP32: Tensor[T], biasFP32: Tensor[T]): this.type = {
+    if (biasFP32 != null) {
+      bias.copy(biasFP32)
+    } else {
+      bias.fill(ev.fromType(0)) // TODO bias may be null, at that time, we should not initialize it
+    }
+
+    val weightFP32Tmp = weightFP32.view(Array(outputSize, inputSize))
+    setWeightSum(weightFP32Tmp, weightSum)
+
+    for (i <- 1 to outputSize) {
+      val singleRow = weightFP32Tmp.select(1, i)
+      min(i - 1) = singleRow.min()
+      max(i - 1) = singleRow.max()
+    }
+
+    val bufferOffset = 0
+    val buffer = new Array[Byte](weightFP32.nElement())
+    val weightFP32Tensor = weightFP32.asInstanceOf[Tensor[Float]]
+    Quant.quantize(weightFP32Tensor, buffer, bufferOffset)
+
+    weight = QuantTensor[T](outputSize, inputSize)
+    weight.setStorage(buffer)
+
+    init()
+
+    this
+  }
+
+  def init(): this.type = {
+    val byteArrayOfWeight = weight.getStorage
+
+    weight.setStorageInJni(
+      BigQuant.FCKernelDescInit(outputSize, inputSize))
+
+    ev.getType() match {
+      case FloatType =>
+        val minArray = min.asInstanceOf[Array[Float]]
+        val maxArray = max.asInstanceOf[Array[Float]]
+
+        BigQuant.FCKernelLoadFromModel(weight.getStorageInJni, byteArrayOfWeight,
+          minArray, maxArray, outputSize, inputSize, WEIGHT_THRESHOLD, BigQuant.NCHW)
+      case _ => throw new UnsupportedOperationException(s"Only support Float for quantized model")
+    }
+
+    _init = true
+
+    this
+  }
+
+  @throws(classOf[IOException])
+  private def writeObject(out: ObjectOutputStream): Unit = {
+    out.defaultWriteObject()
+
+    out.writeObject(weight)
+  }
+
+  @throws(classOf[IOException])
+  private def readObject(in: ObjectInputStream): Unit = {
+    in.defaultReadObject()
+
+    weight = in.readObject().asInstanceOf[QuantTensor[T]]
+
+    if (weight.getStorage != null && weight.getStorageInJni == 0L) {
+      init()
+    }
+  }
+
+  def checkAndInit(): Unit = {
+    if (!_init && weight.getStorageInJni == 0L) {
+      init()
+    }
+  }
+
+  override def updateOutput(input: Tensor[T]): Tensor[T] = {
+    require(input.dim() == 1 || input.dim() == 2,
+      "bigquant.Linear: " + ErrorInfo.constrainInputAsVectorOrBatch)
+
+    checkAndInit()
+
+    val batchSize = if (input.dim() == 1) {
+      output.resize(Array(outputSize)) // TODO
+      1
+    } else {
+      output.resize(Array(input.size(1), outputSize))
+      input.size(1)
+    }
+
+    if (!data.isInitialized) {
+      data.setStorageInJni(BigQuant.FCDataDescInit(batchSize, inputSize))
```

Comment:
What if the batchSize or inputSize changes after data is initialized?

Reply:
No, we need reinitialize it, and no much performance penalty.